include/proxy-wasm/context.h

```diff
@@ -0,0 +1,731 @@
+// Copyright 2016-2019 Envoy Project Authors
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "include/proxy-wasm/compat.h"
+
+#include <time.h>
+#include <atomic>
+#include <chrono>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <vector>
+
+namespace proxy_wasm {
+
+#include "proxy_wasm_common.h"
+#include "proxy_wasm_enums.h"
+
+class WasmBase;
+class WasmVm;
+
+// TODO: use the SDK version when it is updated.
+enum class ProxyAction : uint32_t {
+  Illegal = 0,
+  Continue = 1,
+  Pause = 2,
+};
+
+using Pairs = std::vector<std::pair<string_view, string_view>>;
+using PairsWithStringValues = std::vector<std::pair<string_view, std::string>>;
+using CallOnThreadFunction = std::function<void(std::function<void()>)>;
+
+/**
+ * BufferInterface provides a interface between proxy-specifc buffers and the proxy-independent ABI
+ * implementation. Embedders should subclass BufferInterface to enable the proxy-independent code to
+ * implement ABI calls which use buffers (e.g. the HTTP body).
+ */
+struct BufferInterface {
+  virtual ~BufferInterface() {}
+  virtual size_t size() const = 0;
+  /**
+   * Copy bytes from the buffer into the Wasm VM corresponding to 'wasm'.
+   * @param start is the first buffer byte to copy.
+   * @param length is the length of sequence of buffer bytes to copy.
+   * @param ptr_ptr is the location in the VM address space to place the address of the newly
+   * allocated memory block which contains the copied bytes.
+   * @param size_ptr is the location in the VM address space to place the size of the newly
+   * allocated memory block which contains the copied bytes (e.g. length).
```

Comment:
Is this needed because the real length might be less than the requested length? Can you clarify?

Reply:
The size_ptr is the location in the VM to write the length.   This is just for convenience since otherwise this functionality would require 2 or more calls, to allocate the data, do the copy, write the pointer to the VM , write the size to the VM.

It says (e.g. length... changed to i.e. length) to make it clear that it is length.